AS7-4130  AS 7 quickstart: Wicket (EAR part)

OndraZizka opened the pull request.

OndraZizka committed:
AS7-4130  AS 7 quickstart: Wicket (EAR)

OndraZizka committed:
AS7-4130  Rename packages to match other QS.

OndraZizka committed:
AS7-4130  War part fix.

OndraZizka committed:
AS7-4130  War part persistence.xml fix.

sgilda commented:
Hi Ondrej,
I must admit I am totally confused by this directory structure. This is what I see:
- wicket
  -- ear-parent
     --- ear
     --- ejb
     --- war
   -- war
  There is no README.md file in the root directory to explain what the quickstart is or how to deploy and run it. I do see a README in the higher level "war" directory, but it seems to be a copy of the one from the "ejb-in-war" quickstart. I also see a README.txt file in the "ear-parent" directory, but it doesn't tell me much. I'm unclear how to proceed with this one.

Also, in the pom.xml file, the groupId should be "org.jboss.as.quickstarts" and the artifactId shoul match the standard naming of  "jboss-as-wicket". See https://docs.jboss.org/author/display/AS71/Maven+POM+Versions+Checklist and https://docs.jboss.org/author/display/AS71/Writing+a+quickstart  for details.  You will need to update the subdirectory pom.xml file also. 

Thanks,
Sande

OndraZizka commented:
Hi Sande, thanks for the feedback.
The dir structure reflects the purpose of the example. I'll write that in README.md.

pmuir commented:
This looks like two quickstarts. One for Wicket in an EAR, and one for Wicket in a war? If so, it should be two quickstarts, with wicket-ear and wicket-war as the names. Can you sort that out, and then send two pulls, one for each. If the EAR is based on the war or vs versa, then it would be sensible to get one reviewed first, so the other can be updated and submitted. I suggest starting with the war if there is no difference at your end...

OndraZizka committed:
Add JBoss Java EE BOM.

OndraZizka committed:
Move Wicket EAR and WAR qs to top level dir.

OndraZizka committed:
AS7-4130, conform to quickstarts guidelines https://docs.jboss.org/author/display/AS71/Writing+a+quickstart

OndraZizka committed:
Remove legacy README.txt

OndraZizka commented:
Updated:
1) Split to wicket-ear and wicket-war
2) Went through the guidelines and applied what was applicable
3) Rebased
4) Created README.md for both

Tested, works.
I hope I didn't miss something important.
Can you please review?

OndraZizka committed:
AS7-4130, conform to quickstarts guidelines

pmuir commented on wicket-ear/README.md:
```diff
@@ -0,0 +1,73 @@
+
+ * Injection of @PersistenceContext
+ * Injection of a value from web.xml using @Resource
+ * Injection of a stateless session bean using @EJB
```
You should enclose inline code with backticks e.g. `@PersistenceContext` or `web.xml`

pmuir commented on wicket-ear/README.md:
```diff
@@ -0,0 +1,73 @@
+* wicket-ear - parent module.
+** ejb: Contains EJB beans and JPA entities. Creates a .jar file.
+** war: Contains the Wicket web application, which uses the EJB beans. Creates a .war file.
+** ear: Packages the EJB JAR and WAR into an EAR. Creates an .ear file.
```
I think each module name should be enclosed in backticks

pmuir commented on wicket-ear/ear/pom.xml:
```diff
@@ -0,0 +1,76 @@
+    <modelVersion>4.0.0</modelVersion>
+    
+    <parent>
+        <groupId>org.jboss.as.quickstart.wicket.ear</groupId>
```
Ithink we should use a non-nested group id such as wicket_ear or wicket-ear - just for consistency.

OndraZizka replied:
Fixed

pmuir commented on wicket-ear/ear/pom.xml:
```diff
@@ -0,0 +1,76 @@
+                <artifactId>maven-ear-plugin</artifactId>
+                <version>2.6</version>
+                <configuration>
+                    <!-- Change this to 5 if you want to use with Java EE 5 container -->
```
nice comment, but i would actually prefer we don't put it here - these quickstarts target JBoss AS 7/ EAP 6, which is an EE 6 container :-)

OndraZizka replied:
Removed

pmuir commented on wicket-ear/ear/pom.xml:
```diff
@@ -0,0 +1,76 @@
+                    <modules>
+                        <webModule>
+                            <groupId>${project.groupId}</groupId>
+                            <artifactId    >as7-quickstart-wicket-ear-war</artifactId>
```
random whitespace here.

OndraZizka replied:
Well, it's not random - it keeps the two lines aligned:

```
                        <artifactId    >as7-quickstart-wicket-ear-war</artifactId>
                        <bundleFileName>as7-quickstart-wicket-ear.war</bundleFileName>
```

To make it obvious that the values are almost the same. IMEx, prevents typos. Personally, I prefer that. U no? :)

sgilda replied:
Couldn't you do that better with a comment?

```
                         <artifactId>as7-quickstart-wicket-ear-war</artifactId>
                         <!-- NOTE: The following bundleFileName value matches the artifactId value above -->
                         <bundleFileName>as7-quickstart-wicket-ear.war</bundleFileName>
```

On 03/23/2012 04:26 AM, Ondrej Zizka wrote:

> > +
> > +<build>
> > +<plugins>
> > +
> > +<!-- EAR archive creation. -->
> > +<plugin>
> > +<groupId>org.apache.maven.plugins</groupId>
> > +<artifactId>maven-ear-plugin</artifactId>
> > +<version>2.6</version>
> > +<configuration>
> > +<!-- Change this to 5 if you want to use with Java EE 5 container -->
> > +<version>6</version>
> > +<modules>
> > +<webModule>
> > +<groupId>${project.groupId}</groupId>
> > +<artifactId>as7-quickstart-wicket-ear-war</artifactId>
> > Well, it's not random - it keeps the two lines aligned:
> 
> ```
>                          <artifactId>as7-quickstart-wicket-ear-war</artifactId>
>                          <bundleFileName>as7-quickstart-wicket-ear.war</bundleFileName>
> ```
> 
> To make it obvious that the value is the same. Personally, I prefer that. U no? :)
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jbossas/quickstart/pull/151/files#r595080

pmuir replied:
Actually looking again, I don't think you need to include the `<bundleFileName>` at all. If you remove it, it will solve the problem.

OndraZizka replied:
It would create war with name like as7-quickstart-wicket-ear-1.0.0.war, which is then used as JNDI indentifier, and would change with every version.

pmuir replied:
Sorry, I'm missing where you do a JNDI lookup?

pmuir commented on wicket-ear/ear/pom.xml:
```diff
@@ -0,0 +1,76 @@
+                        </webModule>
+                        <ejbModule>
+                            <groupId>${project.groupId}</groupId>
+                            <artifactId    >as7-quickstart-wicket-ear-ejbjar</artifactId>
```
Random whitespace here.

OndraZizka replied:
Fixed

pmuir commented on wicket-ear/ear/pom.xml:
```diff
@@ -0,0 +1,76 @@
+            </plugin>
+            
+        </plugins>
+    </build>
```
The build section conventionally goes underneath the dependencies section and above the profile section.

OndraZizka replied:
Never heard of that. Personally I put dependencies down because it tends to be longer than <build>, therefore it's harder to find the start of <build> if it's under.  But... you're the boss :)

pmuir replied:
If you take a look at e.g. the Maven getting started guide, you can see this convention - http://maven.apache.org/guides/getting-started/index.html.

OndraZizka replied:
IMO it's just a coincidence that they put it that way to the example.

pmuir replied:
Regardless, it has become the generally accepted way to do it, in my experience. But the key point is to keep it consistent through the quickstarts.

pmuir commented on wicket-ear/ear/src/main/resources/application.xml:
```diff
@@ -0,0 +1,18 @@
+        </web>
+    </module>
+
+</application>
```
Won't maven generate application.xml for us based on the ear plugin?

pmuir replied:
What about this one?

pmuir commented on wicket-ear/ejb/pom.xml:
```diff
@@ -0,0 +1,61 @@
```
Indent is wrong on this file - should be 4 spaces as for other files in this commit

pmuir replied:
Also, ordering of eleements doesn't follow conventions

OndraZizka replied:
Now really, what conventions do you refer to?

OndraZizka replied:
Fixed whitespace

pmuir replied:
I explained the conventional ordering of elements above. It looks like it's now fixed :-)

pmuir commented on wicket-ear/ejb/pom.xml:
```diff
@@ -0,0 +1,61 @@
+	<build>
+		<plugins>
+            
+            <!-- EJB archive creation. -->
```
Whitespace out here.

OndraZizka replied:
You mean the new line? It's nicer with, IMO. But fixed.

pmuir replied:
No, I meant the lack on 4 space indent on `<packaging>ejb</packaging>`.

OndraZizka replied:
Fixed

pmuir commented on wicket-ear/ejb/pom.xml:
```diff
@@ -0,0 +1,61 @@
+	</parent>
+    
+    <artifactId>as7-quickstart-wicket-ear-ejbjar</artifactId>
+	<packaging>ejb</packaging>
```
Whitespace out here

OndraZizka replied:
Fixed

pmuir commented on wicket-ear/ejb/pom.xml:
```diff
@@ -0,0 +1,61 @@
+        <version>7.1.1-SNAPSHOT</version>
+	</parent>
+    
+    <artifactId>as7-quickstart-wicket-ear-ejbjar</artifactId>
```
The convention we've established is not to use `ejbjar` but `ejb` here.

OndraZizka replied:
Fixed

pmuir commented on wicket-ear/ejb/pom.xml:
```diff
@@ -0,0 +1,61 @@
+					<ejbVersion>3.1</ejbVersion>
+					<archive>
+						<manifest>
+							<addClasspath>true</addClasspath>
```
Are you sure this is right? It needs a comment, regardless

OndraZizka replied:
No, it was probably dragged through versions of the original archetype. Removed.

pmuir commented on wicket-ear/ejb/pom.xml:
```diff
@@ -0,0 +1,61 @@
+    
+	<dependencies>
+        
+        <!-- Java EE 6 API. -->
```
whitespace

OndraZizka replied:
Fixed.

pmuir commented on wicket-ear/ejb/pom.xml:
```diff
@@ -0,0 +1,61 @@
+			<artifactId>javaee-api</artifactId>
+			<version>6.0</version>
+			<scope>provided</scope>
+		</dependency>
```
We use the org.jboss.spec APIs not the ones from Sun. Please take a look at other quickstarts such as kitchensink for examples of how to declare and manage dependencies.

OndraZizka replied:
Ok. What's the difference between org.jboss.bom and org.jboss.spec? Is the latter intended to be used as well for importing?

OndraZizka replied:
A note on this: There seems to be a mismatch in kitchensink's versions in a comment:

```
  <jboss.bom.version>1.0.0.M5</jboss.bom.version>
  <!-- Alternatively, use the line below to use version 3.0.0.Beta1-redhat-1 which is a release certified 
  to work with JBoss EAP 6. It requires you have access to the JBoss EAP 6 maven repository. -->
  <!-- 
  <jboss.bom.version>1.0.0.M5-redhat-1</jboss.bom.version>> 
```

pmuir replied:
Thanks, I've fixed that.

`org.jboss.spec` should be used when you require pure Java EE. If you also require other things (e.g. Arquillian, JBoss Transactions, Hibernate) you will want something from `org.jboss.bom`

OndraZizka replied:
Fixed.

pmuir commented on wicket-ear/ejb/src/main/java/org/jboss/as/quickstarts/wicket/ear/ejbjar/dao/ContactDaoBean.java:
```diff
@@ -0,0 +1,76 @@
```
Needs 4 space indent.v This needs to be done generally, I'm not going to flag it anymore. Also the braces style is wrong. Run the whole thing a style formatter.

OndraZizka replied:
Done generally. Please review if my IDE formatted properly.

pmuir replied:
Yes, looking better. I'll do a full format when I merge as well :-)

pmuir commented on wicket-ear/ejb/src/main/java/org/jboss/as/quickstarts/wicket/ear/ejbjar/dao/ContactDaoBean.java:
```diff
@@ -0,0 +1,76 @@
+	@PersistenceContext
+	private EntityManager em;
+
+	/**
```
can we remove this @InheritDoc crap, it's not standard.

OndraZizka replied:
Removed.

pmuir commented on wicket-ear/ejb/src/main/java/org/jboss/as/quickstarts/wicket/ear/ejbjar/dao/ContactDaoBean.java:
```diff
@@ -0,0 +1,76 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.as.quickstarts.wicket.ear.ejbjar.dao;
```
The package name shouldn't be nested like this - wicket_ear would be conventional.

OndraZizka replied:
Hmm... Let me frown here... Conventional according to what conventions? I personally have  never seen _ in package names, except for my own code :)  And some archaic projects. how about wicketear or wicketEar? I've not seen the latter either though :)

pmuir replied:
Yes, that would be fine as well (I prefer `wicketear`). I really should have said we don't want the nested package.

pmuir commented on wicket-ear/ejb/src/main/java/org/jboss/as/quickstarts/wicket/ear/ejbjar/model/Contact.java:
```diff
@@ -0,0 +1,110 @@
+			return false;
+		}
+		final Contact other = (Contact)obj;
+		if (id != other.id && (id == null || !id.equals(other.id)))
```
The synthetic ID shouldn't be part of the object equality. This is a long known anti-pattern.

OndraZizka replied:
What's the best practice? I mean, if the IDs are different and not null or zero, than it's safe to say they are not equal. But that's not applicable to hashCode since nodes with null ID and non-zero ID would be equal but have different hash.
So, only organic properties?

pmuir replied:
You can read more here - https://community.jboss.org/wiki/EqualsAndHashCode - in this case you could make email a unique column, and use that as the natural identifier.

pmuir commented on wicket-ear/war/src/main/java/org/jboss/as/quickstarts/wicket/ear/war/pages/InsertContact.java:
```diff
@@ -0,0 +1,88 @@
+	private String name;
+	private String email;
+    
+	//@EJB(lookup="java:app/ContactDaoBean")
```
Can we remove commented code

OndraZizka replied:
Fixed.

pmuir commented on wicket-ear/war/src/main/java/org/jboss/as/quickstarts/wicket/ear/war/pages/ListContacts.java:
```diff
@@ -0,0 +1,73 @@
+{
+
+	private static final long serialVersionUID = 1L;
+	@EJB(name = "ContactDaoBean")
```
Shouldn't we use CDI injection, not EJB?

OndraZizka replied:
Not sure - people will look for examples of how to use @EJB... maybe we could add another property injected via CDI? Or an alternative page with same functionality but with @Inject?

pmuir replied:
We (JBoss) don't really recommend the use of `@EJB` injection in Java EE 6. Hence my comment, we shouldn't be showing the old, less good, way of doing stuff.

pmuir commented:
It looks like the war is a the same code as the ear, so once we've got the ear right, we can look at the war.

OndraZizka committed:
AS7-4130 Change jboss.entity.manager.factory.jndi.name from myEMF to java:app/myEMF
to avoid PU JNDI name conflicts when both apps are deployed.

OndraZizka committed:
README backticks

OndraZizka committed:
README more backticks

OndraZizka committed:
Use a non-nested group id such as wicket_ear or wicket-ear - just for consistency.

OndraZizka committed:
Remove Java EE 5 comment.

OndraZizka committed:
Move deps above build.

OndraZizka committed:
Fix whitespace

OndraZizka committed:
Fix whitespace

OndraZizka committed:
Change 'ejbjar' to 'ejb'

OndraZizka committed:
pom.xml done right

pmuir commented:
Let us know when you are ready for re-review.

OndraZizka committed:
Fix formatting.

OndraZizka committed:
More formatting fixes.

OndraZizka committed:
Remove <addClasspath>.

OndraZizka commented:
Ok so except for the nested package wicket.ear and @EJB vs. @Inject, I've updated according to comments.
Can you please review and let me know what's left? Thanks.

Also, I was thinking of creating a bit more complex app.
Something with JPA with cca. 3 entities, security, JMS, timers.
Is that suitable for a quickstart?

OndraZizka committed:
Fix formatting.

pmuir commented:
I think that sounds too complex for a quickstart really.

OndraZizka committed:
Package names made non-nested.

OndraZizka committed:
Update Wicket version to 1.5.5

OndraZizka commented:
Rebased, tested. Ready for review.

OndraZizka commented:
BTW, under which license should this be released?

pmuir commented:
All quickstarts should be Apache License 2.0. There was some confusion over this, and we are in the process of moving all quickstarts to Apache License 2.0.

pmuir commented:
Pull request #164 replaces this the ear half of this one. Sande, can you review. Once you're done, I will merge down to the war version and send that in.

pmuir commented:
Pull request #165 replaces the war half of this one. Sande, please review :-)

I'll close this one now.
